src/packages/utils/src/things/json-rpc/json-rpc-base-types.ts

```diff
@@ -1,116 +1,43 @@
 
-  protected value: T;
+  protected bufferValue: any | null;
+
   // used to make console.log debugging a little easier
-  private [inspect](_depth: number, _options: any): T {
-    return this.value;
+  private [inspect](_depth: number, _options: any): string {
+    return `[${this.constructor.name}] ${this.toString()}`;
   }
-  constructor(value: T) {
-    const self = this;
-    if (Buffer.isBuffer(value)) {
-      toStrings.set(this, () => value.toString("hex"));
-      bufCache.set(this, value);
-      self[Symbol.toStringTag] = "Buffer";
-    } else {
-      const type = typeof value;
-      switch (type) {
-        case "number":
-          if ((value as number) % 1 !== 0) {
-            throw new Error("`Cannot wrap a decimal value as a json-rpc type`");
-          }
-          toStrings.set(this, () => (value as number).toString(16));
-          toBuffers.set(this, () =>
-            value === 0 ? BUFFER_EMPTY : uintToBuffer(value as number)
-          );
-          break;
-        case "bigint":
-          toStrings.set(this, () => (value as bigint).toString(16));
-          toBuffers.set(this, () =>
-            value === 0n ? BUFFER_EMPTY : bigIntToBuffer(value as bigint)
-          );
-          break;
-        case "string": {
-          // handle hex-encoded string
-          if ((value as string).indexOf("0x") === 0) {
-            toStrings.set(this, () => (value as string).toLowerCase().slice(2));
-            strCache.set(this, (value as string).toLowerCase());
-            toBuffers.set(this, () => {
-              let fixedValue = (value as string).slice(2);
-              if (fixedValue.length % 2 === 1) {
-                fixedValue = "0" + fixedValue;
-              }
-              return Buffer.from(fixedValue, "hex");
-            });
-          } else {
-            throw new Error(
-              `cannot convert string value "${value}" into type \`${this.constructor.name}\`; strings must be hex-encoded and prefixed with "0x".`
-            );
-          }
-          break;
-        }
-        default:
-          // handle undefined/null
-          if (value == null) {
-            // This is a weird thing that returns undefined/null for a call
-            // to toString().
-            this.toString = () => value as string;
-            bufCache.set(this, BUFFER_EMPTY);
-            break;
-          }
-          throw new Error(`Cannot wrap a "${type}" as a json-rpc type`);
-      }
-      self[Symbol.toStringTag] = type;
-    }
 
-    this.value = value;
+  constructor(value: JsonRpcInputArg) {
+    this[Symbol.toStringTag] = typeof(value);
+
+    const parseAndValidate = getParseAndValidateFor(value);
+    this.bufferValue = parseAndValidate(value);
   }
 
   toString(): string | null {
-    let str = strCache.get(this);
-    if (str === void 0) {
-      str = "0x" + toStrings.get(this)();
-      strCache.set(this, str);
+    if (this.bufferValue == null) {
+      return <null>this.bufferValue;
```

Comment:
Any reason not to do:

```suggestion
      return null;
```

Reply:
The previous implementation explicitly distinguished between null and undefined in at least one case: https://github.com/trufflesuite/ganache/blob/c4be4b7a278d5645c26b4499aecf18d9c2410492/src/packages/utils/src/things/json-rpc/json-rpc-quantity.ts#L75

There's a test that asserts that `nonce === undefined`:

https://github.com/trufflesuite/ganache/blob/c4be4b7a278d5645c26b4499aecf18d9c2410492/src/chains/ethereum/ethereum/tests/transaction-pool.test.ts#L460

Do you have thoughts on whether we _should_ distinguish?

Reply:
I think we should tighten up our types and enable `strictNullChecks` in our tsconfig.json some day Soon:tm:. But today is probably not that day. :-)

In this case maybe we should delete `valueOf`? Other than that test it looks like it is only _explicitly_ used in blockchain.ts's `revert`; and we can probably change that to not need it without much effort.

I originally intended on using `valueOf` to do arithmetic on Quantity types without having to call `toBigInt()` first: e.g.:

```javascript
// `valueOf` magic
Quantity.from("0x5") + Quantity.from(Buffer.from(2)) // result: 7n !
```

Typescript didn't like it though, but the `valueOf` function remains. I'm not sure if there are any implicit usages remaining.

I vote to remove `valueOf` if it's trivial to do so!

Reply:
I created this case https://github.com/trufflesuite/ganache/issues/3169 to cover removing valueOf, as there's at least some implicit usage of the function that make it slightly less trivial to remove.